### PR TITLE
Fixes #33727 - Fixed gathering of allowed actions for permission

### PR DESCRIPTION
**Pull request by domitea**
Plugins can define own permissions. They can also extend existing permissions with own actions (https://github.com/theforeman/foreman_ansible/blob/master/lib/foreman_ansible/register.rb#L149). Or not? I figured out that the definition of existing permission in plugin causes creating another Permission with the same name as the core counterpart but with only plugin-specific actions. Then, when you try to do some action defined by that plugin you end up with the permission error.
It's happen because method `allowed_actions` method look for the permission by `permission` method and it's find only first occurrence (that's the behavior of the `detect` method) of that permission and that's the core one without action defined in plugin and that's never end good.

I'm marking this PR as draft because bad move in roles and permissions can cause a nice security problem but even that the bug is caused by `permission()` method I didn't fixed it there. This method is public and it's used in plugin for extending existing permissions (https://github.com/theforeman/foreman_puppet/blob/master/lib/foreman_puppet/register.rb#L9). 

This topic is complex and I think that's great topic for further discussion because I that foreman deserve better way how to handle extending of existing permissions by plugins but now I just try to fix the bug until we find a better way.

### CODE REVIEW

File: app/registries/foreman/access_control.rb [hunk 1]
```diff
@@ -55,8 +55,8 @@ def remove_permission(permission)
 
       # Returns the actions that are allowed by the permission of given name
       def allowed_actions(permission_name)
-        perm = permission(permission_name)
-        perm ? perm.actions : []
+        perm = permissions.filter { |p| p.name == permission_name }
+        perm ? perm.map(&:actions).flatten.uniq : []
```
Comment (ares) [inline app/registries/foreman/access_control.rb]: I think `perm` will now always be truthy, since it's an array and even empty array is considered truthy, we should probably use `perm.present?`

**Review by ares — COMMENTED** [review]
The tests failures seems to be related, perhaps some edge case when something is not found, since the backend now returns 500 for many tests

**Comment by ezr-ondrej** [comment]
Wouldn't it be easier to uniq the permissions?
In the future I would imagine we should have method to add allowed actions to existing permissions (the foreman_puppet usecase) and just forcefully fail on adding permission that exists.

In the meantime (this PR) I would just say that plugins are trying to extend existing permissions through adding it again, so I'd just add the actions to one permission (it would definitelly be desired state for the Ansible).

File: app/registries/foreman/access_control.rb [hunk 2]
```diff
@@ -55,8 +55,8 @@ def remove_permission(permission)
 
       # Returns the actions that are allowed by the permission of given name
       def allowed_actions(permission_name)
-        perm = permission(permission_name)
-        perm ? perm.actions : []
+        perm = permissions&.filter { |p| p.name == permission_name }
```
Comment (ezr-ondrej) [inline app/registries/foreman/access_control.rb]: `filter` is only ruby > 2.6, that's why it fails on 2.5 thus the test failures.
```suggestion
        perm = permissions.select { |p| p.name == permission_name }
```

**Review by ezr-ondrej — CHANGES_REQUESTED** [review]
Firstly @domitea nice catch finding out where the issue came from, but I do not agree with this solution.

if we allow to have two permissions with the same name this doesn't need to be the only place we see unexpected behaviour, other places are designed to work with only single permission with given name -namely first the DB model: https://github.com/theforeman/foreman/blob/e20468076a67581d951ae1fb5114ea4d2164f1c8/app/models/permission.rb#L3

You've mentioned `security_block`, but it is functionally irrelevant, so there is no reason not to ignore it it is only permission "category", it is better ignoring it with deprecation warning, then having unexpected behaviour.
I'd really love if we could actually avoid the duplications with something like:
```ruby
      def add_permission_actions(perm_name, actions)
         #initialize the Hash somewhere
         @added_permission_actions[perm_name] ||= []
         @added_permission_actions[perm_name].concat(actions)
      end

      def permission(name, hash, options = {})
        @permissions ||= []
        options[:security_block] = @security_block if @security_block
        perm = # find perm Foreman::AccessControl.permission(name)
        if perm
          #DeprecationWarning('It is unsupported to create two permissions with the same name use add_permission_actions instead', 3.2)
          add_permission_actions(perm.name, options[:actions])
        else
          @permissions << Permission.new(name, hash, options)
        end
      end
```

And then obviously merging the actions at the end of `Foreman::AccessControl.map`

File: app/registries/foreman/access_control.rb [hunk 3]
```diff
@@ -23,6 +23,16 @@ def map
...
+        mapper.actions_to_add.each do |perm_name, actions_with_controller|
+          p = permission(perm_name)
+          allowed_actions = []
+          actions_with_controller.map do |controller, actions|
+            actions.each do |action|
+              allowed_actions << "#{controller}/#{action}"
+            end
+          end
+          p.actions.concat(allowed_actions)
+        end
```
Comment (ezr-ondrej) [inline app/registries/foreman/access_control.rb]: Seeing this, wouldn't it be easier to just to initialize the Permission in every case and on the collision just have `p.actions.concat(collision_perm.actions)`?

We could even ignore collisions in the mapper and resolve them here.
```ruby
@permissions ||= []
add_permissions(mapper.mapped_permissions)
add_permissions(mapper.added_permission_actions, true)

def add_permissions(permissions, adding = false)
  permissions.each do |perm|
    if (p = permission(perm.name))
      # Deprecation if adding
      p.actions.concat(perm.actions)
    else
      # error perm missing if !adding
      @permissions << perm
    end
  end
end
```

File: app/registries/foreman/access_control.rb [hunk 4]
```diff
@@ -100,6 +122,10 @@ def security_block(name, options = {})
       def mapped_permissions
         @permissions
       end
+
+      def actions_to_add
+        @added_permission_actions.flatten(0)
```
Comment (ezr-ondrej) [inline app/registries/foreman/access_control.rb]: I needed to read into the docs to find out what this does.
I think this only does `to_a` and I do not get why, we can iterate over the hash directly

**Review by ezr-ondrej — COMMENTED** [review]
I do not like the duplication of the `AccessControl::Permission` code. I know it's comming from following my instructions from previous review, but I did not realize it will result in this, can't we use the `AccessControll::Permission` to do that for us and store it in that structure, just in the hash?

Or I had an idea, that we could even resolve the conflicts once merging the mapper permissions to the global ones and in the mapper store them in a single storage (ignoring the collisions). That has the additional value, that there would not be an additional logic in a `Mapper`, that is supposed to be quite "dumb" class. For this I've shared hinters below, but feel free to take a different direction :)

**Comment by domitea** [comment]
Thanks for comments! I just wrote my first idea about that when I figured out what's problem it is. I'm just merging permission's actions with the same name in the `map` method.

File: app/registries/foreman/access_control.rb [hunk 5]
```diff
@@ -55,8 +61,8 @@ def remove_permission(permission)
 
       # Returns the actions that are allowed by the permission of given name
       def allowed_actions(permission_name)
-        perm = permission(permission_name)
-        perm ? perm.actions : []
+        perm = permissions.select { |p| p.name == permission_name }
+        perm.present? ? perm.map(&:actions).flatten.uniq : []
```
Comment (ezr-ondrej) [inline app/registries/foreman/access_control.rb]: we do not need this, now, right?

File: app/registries/foreman/access_control.rb [hunk 6]
```diff
@@ -22,7 +22,13 @@ def map
         mapper = Mapper.new
         yield mapper
         @permissions ||= []
-        @permissions += mapper.mapped_permissions
+        mapper.mapped_permissions.each do |permission|
+          if (perm = permission(permission.name))
+            perm.actions.concat(permission.actions)
```
Comment (ezr-ondrej) [inline app/registries/foreman/access_control.rb:28]: I think if you just add an warning here, we can add the proper API in another PR and consider this a fix for the bug :)

**Review by ezr-ondrej — COMMENTED** [review]
I think this is good enough, if you add an warning.
Can you add it and undraft this PR?

File: app/registries/foreman/access_control.rb [hunk 7]
```diff
@@ -22,7 +22,14 @@ def map
         mapper = Mapper.new
         yield mapper
         @permissions ||= []
-        @permissions += mapper.mapped_permissions
+        mapper.mapped_permissions.each do |permission|
+          if (perm = permission(permission.name))
+            Rails.logger.warn "You are trying to replace #{perm.name} from #{permission.engine}. Adding actions from plugin permissions to a core one."
```
Comment (ezr-ondrej) [inline app/registries/foreman/access_control.rb]: logger should be enough I think
```suggestion
            logger.warn "You are trying to replace #{perm.name} from #{permission.engine}. Adding allowed actions from plugin permissions to the existing one."
```

**Review by ezr-ondrej — COMMENTED** [review]
Just wording and I think that `logger` is set, and actually we are patching it, so it might be a different logger then Rails.logger (not sure there, but we never use Rails.logger, so better be consistent).

**Comment by domitea** [comment]
Thanks for the wording. I updated that but I really like to stay with `Rails.logger` call because it's used across several registries (https://github.com/theforeman/foreman/blob/develop/app/registries/foreman/plugin.rb#L112). Also alone call of `logger` causes `undefined local variable or method `logger'`

**Comment by ezr-ondrej** [comment]
> Also alone call of `logger` causes `undefined local variable or method `logger'`

Make sense, the global logger is probably not ready yet at that point.
Undrafting for you :)

**Review by ezr-ondrej — APPROVED** [review]
works for me :+1:

**Comment by ezr-ondrej** [comment]
Looks safe, hopefully it won't break anything :)

**Comment by ezr-ondrej** [comment]
CP 3.0: #8890
